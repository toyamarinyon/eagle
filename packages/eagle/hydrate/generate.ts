import { build } from "esbuild";
import { writeFileSync, rmSync } from "fs";
import { basename, dirname, extname, join } from "path";

// @ts-ignore
import { skypackPlugin } from "../build/skypackPlugin.js";

const hydrateCodeTemplate = `
import { hydrate } from "react-dom"
import Page from "{path}"
type PageProps = Record<string, any> | undefined | null

function renderPage(props: PageProps) {
// @ts-ignore
 return hydrate(<Page {...props} />, document.getElementById('eagle-root'))
}

// @ts-ignore
let props: PageProps = {/** placeholder */}

renderPage(props)
`;

export async function renderString(runtimeDir: string, path: string) {
  const ext = extname(path);
  const base = basename(path, ext);
  const dir = dirname(path);
  const tsCode = hydrateCodeTemplate.replace(
    "{path}",
    join("../../../", join(dir, base))
  );
  const tmpPath = join(runtimeDir, "tmp", path.replace(/\//g, "-"));
  writeFileSync(tmpPath, tsCode);
  const buildResult = await build({
    entryPoints: [tmpPath],
    inject: [join(runtimeDir, "reactShim.ts")],
    target: "es2022",
    format: "esm",
    bundle: true,
    plugins: [skypackPlugin],
    write: false,
  });
  rmSync(tmpPath);

  return buildResult.outputFiles[0].text;
}

export async function createHydrateRouter(
  runtimeDir: string,
  filePaths: string[]
) {
  const tmp = await Promise.all(
    filePaths.map(async (filePath) => {
      const hydrateString = await renderString(runtimeDir, filePath);
      const ext = extname(filePath);
      const base = basename(filePath, ext);
      const dir = dirname(filePath);
      const subdir = dir.replace(/^src\/pages\/?/, "");
      return [join(subdir, base), hydrateString];
    })
  );
  return `
  export const hydrateRoutes = {
  ${tmp
    .map(([filePath, hydrateString]) => `["${filePath}"]: \`${hydrateString}\``)
    .join(",\n  ")}
  }`;
}
