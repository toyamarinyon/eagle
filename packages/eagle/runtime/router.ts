import { Page } from "./page";

type Route = () => Promise<Page>;
export type Routes = { [key: string]: Route };

export async function router(request: Request, routes: Routes) {
  const url = new URL(request.url);
  const pathname = pathnameToFilePath(url.pathname);
  if (routes[pathname] == undefined) {
    throw new Error(`Page not found: ${pathname}`);
  }
  const pageFile = routes[pathname];
  const page = await pageFile();
  return page
}

export function pathnameToFilePath(pathname: string): string {
  const filePath = pathname
    .replace(/^\/$/, "index")
    .replace(/\/$/, "/index")
    .replace(/^\//, "");
  return filePath;
}
