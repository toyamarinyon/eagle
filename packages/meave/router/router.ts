import { readdirSync, statSync } from "fs";
import { join, extname, basename, dirname } from "path";

export function getPageFileList(
  dirPath: string,
  arrayOfFiles: string[] = [],
  baseDir = process.cwd()
) {
  const readDirPath = join(baseDir, dirPath);
  const files = readdirSync(readDirPath);

  files.forEach(function (file) {
    const filePath = join(readDirPath, file);
    if (statSync(filePath).isDirectory()) {
      arrayOfFiles = getPageFileList(
        join(dirPath, file),
        arrayOfFiles,
        baseDir
      );
    } else {
      arrayOfFiles.push(join(dirPath, file));
    }
  });

  return arrayOfFiles;
}

const routeTemplate = `["{key}"]: async () => await import("{path}") as unknown as PageFile,`;

export function addRoute(filePath: string) {
  const ext = extname(filePath);
  const base = basename(filePath, ext);
  const dir = dirname(filePath);
  const subdir = dir.replace(/^src\/pages\/?/, "");
  const router = routeTemplate
    .replace("{key}", join(subdir, base))
    .replace("{path}", join("../../", dir, base));
  return router;
}

export function createRouter(filePaths: string[]) {
  return `
export const routes = {
  ${filePaths.map((filePath) => addRoute(filePath)).join("\n  ")}
}
  `;
}
