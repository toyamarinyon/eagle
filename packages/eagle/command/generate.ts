import { join } from "path";
import { existsSync, mkdirSync, writeFileSync } from "fs";
import { createRouter, getPageFileList } from "../router/router";

type GenerateOption = {
  pagesDir: string;
  packageDir: string;
  runtimeDir: string;
};
export function generateRuntime({ pagesDir, packageDir }: GenerateOption) {
  const runtime = `
  import { handler, Page } from "${join(packageDir, "runtime")}";
  ${generateRouter(pagesDir)}
  export async function eagleHandler(request: Request) {
    return await handler(request, routes);
  }
  `;
  return runtime;
}
function generateRouter(pagesDir: string) {
  const files = getPageFileList(pagesDir);
  const router = createRouter(files);
  return router;
}

export function writeRuntime({
  pagesDir = "src/pages",
  packageDir = "@toyamarinyon/eagle",
  runtimeDir = "node_modules/.eagle",
}: Partial<GenerateOption>) {
  const runtime = generateRuntime({ pagesDir, packageDir, runtimeDir });
  if (!existsSync(runtimeDir)) {
    mkdirSync(runtimeDir);
  }
  writeFileSync(join(runtimeDir, "index.ts"), runtime);
}
