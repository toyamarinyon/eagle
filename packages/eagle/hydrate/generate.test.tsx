import { join } from "path";
import { test, expect } from "vitest";
import { getPageFileList } from "../router/router";
import { createHydrateRouter, renderString } from "./generate";

test("rendering hydrate script as string", async () => {
  const stubDir = join(__dirname, "__stub__");
  const pagePath = join("src", "pages", "index.tsx");
  const tmpRuntimeDir = join(stubDir, "node_modules", "runtime");
  const code = await renderString(tmpRuntimeDir, pagePath);
  expect(code).toMatchInlineSnapshot(`
    "// hydrate/__stub__/node_modules/runtime/reactShim.ts
    import * as React from \\"https://cdn.skypack.dev/react\\";

    // hydrate/__stub__/node_modules/runtime/tmp/src-pages-index.tsx
    import { hydrate } from \\"https://cdn.skypack.dev/react-dom\\";

    // hydrate/__stub__/src/pages/index.tsx
    function HelloWorld() {
      return /* @__PURE__ */ React.createElement(\\"div\\", null, \\"hello\\");
    }

    // hydrate/__stub__/node_modules/runtime/tmp/src-pages-index.tsx
    function renderPage(props2) {
      return hydrate(/* @__PURE__ */ React.createElement(HelloWorld, {
        ...props2
      }), document.getElementById(\\"eagle-root\\"));
    }
    var props = {};
    renderPage(props);
    "
  `);
});

test("create hydrate script router", async () => {
  const stubDir = join(__dirname, "__stub__");
  const tmpRuntimeDir = join(stubDir, "node_modules","runtime");
  const filePaths = getPageFileList("src/pages", [], stubDir);
  expect(filePaths).toMatchInlineSnapshot(`
    [
      "src/pages/directory/hello.tsx",
      "src/pages/index.tsx",
    ]
  `);
  const router = await createHydrateRouter(tmpRuntimeDir, filePaths);
  expect(router).toMatchSnapshot();
});
