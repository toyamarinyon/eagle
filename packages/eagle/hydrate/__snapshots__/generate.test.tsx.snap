// Vitest Snapshot v1

exports[`create hydrate script router 2`] = `
"
  export const hydrateRoutes = {
  [\\"directory/hello\\"]: \`// hydrate/__stub__/node_modules/runtime/reactShim.ts
import * as React from \\"https://cdn.skypack.dev/react\\";

// hydrate/__stub__/node_modules/runtime/tmp/src-pages-directory-hello.tsx
import { hydrate } from \\"https://cdn.skypack.dev/react-dom\\";

// hydrate/__stub__/src/pages/directory/hello.tsx
function HelloWorld() {
  return <h1>Hello World</h1>;
}

// hydrate/__stub__/node_modules/runtime/tmp/src-pages-directory-hello.tsx
function renderPage(props2) {
  return hydrate(/* @__PURE__ */ React.createElement(HelloWorld, {
    ...props2
  }), document.getElementById(\\"eagle-root\\"));
}
var props = {};
renderPage(props);
\`,
  [\\"index\\"]: \`// hydrate/__stub__/node_modules/runtime/reactShim.ts
import * as React from \\"https://cdn.skypack.dev/react\\";

// hydrate/__stub__/node_modules/runtime/tmp/src-pages-index.tsx
import { hydrate } from \\"https://cdn.skypack.dev/react-dom\\";

// hydrate/__stub__/src/pages/index.tsx
var test = \\"hello\\";
function HelloWorld() {
  const testMessage = \\\\\`
  \\\\\${test}\\\\\`;
  return <div>{testMessage}</div>;
}

// hydrate/__stub__/node_modules/runtime/tmp/src-pages-index.tsx
function renderPage(props2) {
  return hydrate(/* @__PURE__ */ React.createElement(HelloWorld, {
    ...props2
  }), document.getElementById(\\"eagle-root\\"));
}
var props = {};
renderPage(props);
\`
  }"
`;
